homeassistant/components/decora_wifi/__init__.py

```diff
@@ -78,34 +83,47 @@ async def async_setup_entry(hass: HomeAssistant, entry):
             )
         }
     )
-    devices = {}
-
-    try:
-        devices = await DecoraWifiPlatform.async_getdevices(hass, email)
-    except DecoraWifiCommFailed:
-        _LOGGER.error("Communication with Decora Wifi platform failed.")
-        return False
-    except DecoraWifiSessionNotFound:
-        # Re-login.
-        if await DecoraWifiPlatform.async_login(hass, email, password):
-            devices = await DecoraWifiPlatform.async_getdevices(hass, email)
-
-    # Forward the config entry for each device type present.
-    for p in PLATFORMS:
-        if devices[p]:
-            hass.async_create_task(
-                hass.config_entries.async_forward_entry_setup(entry, p)
+
+    # If a session is already setup, don't start a new one.
+    session = decorawifisessions.get(email, None)
+    if not session:
+        # Re-login
+        try:
+            # Attempt to log in.
+            session = await DecoraWifiPlatform.async_setup_decora_wifi(
+                hass,
+                email,
+                password,
             )
-    return True
+        except DecoraWifiLoginFailed:
+            _LOGGER.error("Login failed.")
+        except DecoraWifiCommFailed:
+            _LOGGER.error("Communication with myLeviton failed.")
+        decorawifisessions.update({email: session})
+
+    if session:
```

Comment:
This second `if` shouldn't be necessary, but mypy complained because it still considered it possible for `session` to be `None`.